### PR TITLE
Support for accessing Azure repositories through a proxy

### LINKED ISSUE #23506: Support for accessing Azure repositories through a proxy
Consider adding support for accessing an Azure snapshot repository through a proxy. At the moment, the only closely related option is to use `https.proxyPort/Host` java options, which are ignored for this type of Azure access.

### PR DESCRIPTION
**Pull request by dadoonet**
You can define a proxy using the following settings:

```yml
azure.client.default.proxy.host: proxy.host
azure.client.default.proxy.port: 8888
azure.client.default.proxy.type: http
```

Supported values for `proxy.type` are `http` or `socks`. Defaults to no proxy.

Note that this commit depends on #23517 which needs to be merged first (upgrade azure client to 5.0.0).

Closes #23506

### CODE REVIEW

**Comment by dadoonet** [comment]
@tlrx Once you have reviewed #23517, could you review this one please?

**Comment by dadoonet** [comment]
I'm planning to merge this on friday, March 31st, unless anyone objects.

**Comment by jasontedor** [comment]
> I'm planning to merge this on friday, March 31st, unless anyone objects.

I object to this, this requires a review.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (danielmitterdorfer) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: nit: every field assignment uses `this` except for `proxy`. I though initially that it's a local variable. Maybe use `this.proxy` instead of just `proxy`?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (danielmitterdorfer) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Nit: copy & paste error: "The host name of a proxy to connect to **azure** through." (also: shouldn't it be "Azure" instead of "azure"?)

**Review by danielmitterdorfer — COMMENTED** [review]
I left a couple of very minor comments. However, I think it'd be good to add a test?

File: docs/plugins/repository-azure.asciidoc
Comment (tlrx) [inline docs/plugins/repository-azure.asciidoc]: I don't think we need to document the proxy.type `direct`, because it actually does nothing on Azure client side. We could instead check that proxy.type must be either http or socks and require non null host and port if type is set? Similarly, if host (or port) is set, port (or host) and type are required?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (tlrx) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: I think we need to check that proxyHost is non null? Otherwise it's going to use 0.0.0.0 as address.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (tlrx) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Maximum value allowed here is 65536 but later in the code `new InetSocketAddress(proxyHost, proxyPort))` checks that _A valid port value is between 0 and 65535_. Max value here should be 65535.

**Review by tlrx — CHANGES_REQUESTED** [review]
Sorry @dadoonet, this review request ran out of my scope.

I left some comments, nothing big. I'd like a test for this too as @danielmitterdorfer already suggested.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Why is this suppression here? The "reason" doesn't really explain anything. I assume it is because of call to construction of `InetSocketAddress` below. But this is not a valid reason to suppress; the message with forbidden apis clearly states you should look up the address of the host yourself with InetAddress, and then use that ctor for `InetSocketAddress`.

Comment (dadoonet) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Makes sense. Should I simply throw the `UnknownHostException` here or catch it and wrap it into another Exception?

Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: I would just let the `UnknownHostException` bubble up. Wrapping just creates longer stacktraces that are more difficult to read.

**Comment by clintongormley** [comment]
@dadoonet are you planning on coming back to this?

**Comment by dadoonet** [comment]
Definitely yes

**Comment by dadoonet** [comment]
While I was working on it, I realized that you can actually define only a proxy globally and not per client instance which is not a big deal IMO.

But then it was not making sense anymore to define proxy settings per client. So I changed it and now you define proxy settings globally whatever the credentials you are going to use:

```yml
cloud.azure.storage.proxy:
    host: proxy.host
    port: 8888
    type: http
```

I updated the description of this PR accordingly.

I added some tests about proxy settings configuration but I did not add any integration test which uses a Proxy behind the scene.
If you want me to implement such a thing, can you give me some pointers if we already did something like this in the code (simulating a Proxy in IT).

Thanks!

**Comment by rjernst** [comment]
@dadoonet Is this something you still want to get in? I'm a little hesistant of having these proxy settings differ from those in repository-s3 (which are per client). Also, what about secure proxies? For repository-s3 we have the username/password as secure settings for the proxy.

**Comment by dadoonet** [comment]
I'd love to get another review of the recent changes I pushed. And yes, I'd like to add this feature if possible.

> I'm a little hesistant of having these proxy settings differ from those in repository-s3 (which are per client). 

I can ask the Microsoft SDK team if there is a way to pass the proxy per request or per client instead of a global setting.

> Also, what about secure proxies? For repository-s3 we have the username/password as secure settings for the proxy.

I started "small". I don't mean to cover all the use cases first. If this PR gets merged, then I can work on secured proxy later on.

**Comment by dadoonet** [comment]
According to the discussion here: https://github.com/Azure/azure-storage-java/issues/48

We can't pass a proxy setting to the client but we can change the OperationContext for each request apparently.
It would be calling `OperationContext.setProxy(proxy_instance_per_client)` instead of calling today `OperationContext.setDefaultProxy(global_proxy)`.

WDYT @rjernst. Wanna me to implement that instead?

Although I' not sure there is a real use case of having multiple Azure clients with different proxy settings within the same cluster.

**Comment by dadoonet** [comment]
@rjernst Ping. WDYT? Wanna me to try to implement what I proposed at https://github.com/elastic/elasticsearch/pull/23518#issuecomment-312283782?

**Comment by rjernst** [comment]
@dadoonet I'm sorry for the delayed reply. I think your suggestion of using `OperationContext.setProxy` sounds like something reasonable to try in order to achieve per client proxies to match those we have in repository-s3. Do you want to try that and bring this PR up to date?

**Comment by dadoonet** [comment]
@rjernst I pushed a new change. I think it's definitely better than the global setting. BTW the new implementation will allow us to have a finer control to the Azure API calls if needed in the future as we now call `azureSdk.performXXX(null /* accessCondition */, /* options */, operationContext)` instead of `azureSdk.performXXX()` and the operationContext is now build from the azure Settings (`generateOperationContext(clientName)` method).

BTW as a next step not related to this change, I'd like to squash all the packages we have in one single `o.e.repositories.azure`.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: nit: space after `"account",`

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java
Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: `direct` is not one of the values in the docs added for proxy.type in this PR?

**Review by rjernst — APPROVED** [review]
LGTM, thanks @dadoonet

**Comment by dadoonet** [comment]
Thanks @rjernst! Pushed in master and 6.x.

**Pull request merged**
